### DISCUSSION TITLE
Handling links to missing docs/sections

### DISCUSSION THREAD

**Discussion by edrex** [artempyanykh/marksman#119] [Ideas]
Currently links to unresolvable docs/sections can't be used, wiki-style, to create the missing resource. For me at least, this ability is an important aspect of editor support for markdown document stores.

How could this need be served in the context of a language server?

From a UX perspective, simply following the link (go to definition) would ideally cause the missing file/section to be created. If the server points to a non-existent file, can we rely on the editor to handle that? If so, maybe there could be a "wiki mode" where references always resolve, and following links opens a buffer on the new file. For headings, it could just point to the bottom or top of the file and rely on the user to add the correct heading.

For non-wiki mode maybe a code action could be used to create the missing file/heading.
(ps I started writing this as an aside in a comment on #117, but didn't want to hijack that PR.)

**Comment by artempyanykh** [comment]
Thanks for raising this @edrex! As I see, the LSP protocol doesn't allow doing side-effects during goto-def resolution. However, I think this is something that can be done with a code action.

E.g. in the following situation:
```
Link to [[non-existent]] document.
```

When the cursor is over the `[[non-existent]]` the server can suggest a quickfix code action 'Create a missing document' which would create a file. It would require some configuration as to what should be the filename, etc; but this could be done. The only challenging part is that code actions don't involve any user interaction, hence the emphasis on *configuration*.

If you have some other use-cases or specific ideas as to what kind of error -> quickfix, please do share!

**Reply by edrex** [reply]
Code actions to create a missing link target (file or heading) seem sensible, and would at least provide a way to support forward-traversal. Headings would have to be inserted at the bottom of the file, which would require manual editing.

I'm curious how editors would handle the LS pointing to a non-existent file. I'm guessing at least some would open a new buffer at that location. If this worked, it would provide an idea way to handle missing file (visit the location without writing) and also a way to handle a missing heading (just point to the file, and let the user insert the heading). But I don't know anything about LSP, so maybe there's something in the interaction model of the protocol that precludes this outright.

**Reply by megalithic** [reply]
Has there be any particular movement on this by anyone? Came here looking for similar solutions.

**Reply by artempyanykh** [reply]
@megalithic a code action to create a file for a non-existent wiki link has been implemented. Nothing around missing headings though.
